File: widgetssdk/src/main/java/com/glia/widgets/core/dialog/domain/IsShowOverlayPermissionRequestDialogUseCase.java
package com.glia.widgets.core.dialog.domain;

import com.glia.widgets.core.configuration.GliaSdkConfigurationManager;
import com.glia.widgets.core.dialog.PermissionDialogManager;
import com.glia.widgets.core.permissions.PermissionManager;

public class IsShowOverlayPermissionRequestDialogUseCase {
    private final PermissionDialogManager permissionDialogManager;
    private final PermissionManager permissionManager;
    private final GliaSdkConfigurationManager gliaSdkConfigurationManager;

    public IsShowOverlayPermissionRequestDialogUseCase(
            PermissionManager permissionManager,
            PermissionDialogManager permissionDialogManager,
            GliaSdkConfigurationManager gliaSdkConfigurationManager
    ) {
        this.permissionManager = permissionManager;
        this.permissionDialogManager = permissionDialogManager;
        this.gliaSdkConfigurationManager = gliaSdkConfigurationManager;
    }

    public boolean execute() {
        return hasNoOverlayPermissions() &&
                hasNotShownOverlayPermissionRequest() &&
                isUseOverlay();
    }

    private boolean hasNoOverlayPermissions() {
        return !permissionManager.hasOverlayPermission();
    }

    private boolean hasNotShownOverlayPermissionRequest() {
        return !permissionDialogManager.hasOverlayPermissionDialogShown();
    }

    private boolean isUseOverlay() {
        return gliaSdkConfigurationManager.isUseOverlay();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/core/dialog/domain/IsShowOverlayPermissionRequestDialogUseCase.kt
package com.glia.widgets.core.dialog.domain

import com.glia.widgets.core.dialog.PermissionDialogManager
import com.glia.widgets.core.permissions.PermissionManager
import com.glia.widgets.launcher.ConfigurationManager

internal interface IsShowOverlayPermissionRequestDialogUseCase {
    operator fun invoke(): Boolean
}

internal class IsShowOverlayPermissionRequestDialogUseCaseImpl(
    private val permissionManager: PermissionManager,
    private val permissionDialogManager: PermissionDialogManager,
    private val gliaSdkConfigurationManager: ConfigurationManager
) : IsShowOverlayPermissionRequestDialogUseCase {
    private val hasNoOverlayPermissions: Boolean get() = !permissionManager.hasOverlayPermission()
    private val hasNotShownOverlayPermissionRequest: Boolean get() = !permissionDialogManager.hasOverlayPermissionDialogShown()
    private val isUseOverlay: Boolean get() = gliaSdkConfigurationManager.enableBubbleOutsideApp

    override fun invoke(): Boolean = hasNoOverlayPermissions && hasNotShownOverlayPermissionRequest && isUseOverlay
}
